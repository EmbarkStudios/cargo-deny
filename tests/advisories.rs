use cargo_deny::{
    advisories::{self, cfg},
    Krates,
};
#[macro_use]
mod utils;

struct TestCtx {
    dbs: advisories::DbSet,
    lock: advisories::PrunedLockfile,
    krates: Krates,
}

fn load() -> TestCtx {
    let md: krates::cm::Metadata = serde_json::from_str(
        &std::fs::read_to_string("tests/test_data/advisories/06_advisories.json").unwrap(),
    )
    .unwrap();

    let krates: Krates = krates::Builder::new()
        .build_with_metadata(md, krates::NoneFilter)
        .unwrap();

    let lock = advisories::load_lockfile(&std::path::Path::new(
        "tests/test_data/advisories/06_Cargo.lock",
    ))
    .unwrap();

    let db = {
        let tmp = tempfile::tempdir().unwrap();
        advisories::DbSet::load(Some(tmp), vec![], advisories::Fetch::Allow).unwrap()
    };

    let lockfile = advisories::PrunedLockfile::prune(lock, &krates);

    TestCtx {
        dbs: db,
        lock: lockfile,
        krates,
    }
}

#[test]
#[ignore]
fn detects_vulnerabilities() {
    let TestCtx { dbs, lock, krates } = load();

    let cfg = "vulnerability = 'deny'";

    let diags = utils::gather_diagnostics::<cfg::Config, _, _>(
        krates,
        "detects_vulnerabilities",
        Some(cfg),
        None,
        |ctx, tx| {
            advisories::check(
                ctx,
                &dbs,
                lock,
                Option::<advisories::NoneReporter>::None,
                tx,
            );
        },
    )
    .unwrap();

    let vuln_diag = diags
        .iter()
        .find(|v| v.pointer("/fields/code") == Some(&serde_json::json!("RUSTSEC-2019-0001")))
        .unwrap();

    assert_field_eq!(vuln_diag, "/fields/severity", "error");
    assert_field_eq!(
        vuln_diag,
        "/fields/message",
        "Uncontrolled recursion leads to abort in HTML serialization"
    );
    assert_field_eq!(
        vuln_diag,
        "/fields/labels/0/message",
        "security vulnerability detected"
    );
}

#[test]
#[ignore]
fn detects_unmaintained() {
    let TestCtx { dbs, lock, krates } = load();

    let cfg = "unmaintained = 'warn'";

    let diags = utils::gather_diagnostics::<cfg::Config, _, _>(
        krates,
        "detects_unmaintained",
        Some(cfg),
        None,
        |ctx, tx| {
            advisories::check(
                ctx,
                &dbs,
                lock,
                Option::<advisories::NoneReporter>::None,
                tx,
            );
        },
    )
    .unwrap();

    let unmaintained_diag = diags
        .iter()
        .find(|v| v.pointer("/fields/code") == Some(&serde_json::json!("RUSTSEC-2016-0004")))
        .unwrap();

    assert_field_eq!(unmaintained_diag, "/fields/severity", "warning");
    assert_field_eq!(
        unmaintained_diag,
        "/fields/message",
        "libusb is unmaintained; use rusb instead"
    );
    assert_field_eq!(
        unmaintained_diag,
        "/fields/labels/0/message",
        "unmaintained advisory detected"
    );
}

#[test]
#[ignore]
fn detects_unsound() {
    let TestCtx { dbs, lock, krates } = load();

    let cfg = "unsound = 'warn'";

    let diags = utils::gather_diagnostics::<cfg::Config, _, _>(
        krates,
        "detects_unsound",
        Some(cfg),
        None,
        |ctx, tx| {
            advisories::check(
                ctx,
                &dbs,
                lock,
                Option::<advisories::NoneReporter>::None,
                tx,
            );
        },
    )
    .unwrap();

    let unsound_diag = diags
        .iter()
        .find(|v| v.pointer("/fields/code") == Some(&serde_json::json!("RUSTSEC-2019-0035")))
        .unwrap();

    assert_field_eq!(unsound_diag, "/fields/severity", "warning");
    assert_field_eq!(unsound_diag, "/fields/message", "Unaligned memory access");
    assert_field_eq!(
        unsound_diag,
        "/fields/labels/0/message",
        "unsound advisory detected"
    );
    assert_field_eq!(
        unsound_diag,
        "/fields/labels/0/span",
        "rand_core 0.3.1 registry+https://github.com/rust-lang/crates.io-index"
    );
}

#[test]
#[ignore]
fn downgrades_lint_levels() {
    let TestCtx { dbs, lock, krates } = load();

    let cfg = "unmaintained = 'warn'
    ignore = ['RUSTSEC-2016-0004', 'RUSTSEC-2019-0001']";

    let diags = utils::gather_diagnostics::<cfg::Config, _, _>(
        krates,
        "downgrades_lint_levels",
        Some(cfg),
        None,
        |ctx, tx| {
            advisories::check(
                ctx,
                &dbs,
                lock,
                Option::<advisories::NoneReporter>::None,
                tx,
            );
        },
    )
    .unwrap();

    assert_field_eq!(
        diags
            .iter()
            .find(|v| field_eq!(v, "/fields/code", "RUSTSEC-2016-0004"))
            .unwrap(),
        "/fields/severity",
        "help"
    );

    assert_field_eq!(
        diags
            .iter()
            .find(|v| field_eq!(v, "/fields/code", "RUSTSEC-2019-0001"))
            .unwrap(),
        "/fields/severity",
        "help"
    );
}

#[test]
#[ignore]
fn detects_yanked() {
    // Force fetch the index just in case
    rustsec::registry::Index::fetch().unwrap();

    let TestCtx { dbs, lock, krates } = load();

    let cfg = "yanked = 'deny'
    unmaintained = 'allow'
    vulnerability = 'allow'
    ";

    let diags = utils::gather_diagnostics::<cfg::Config, _, _>(
        krates,
        "detects_yanked",
        Some(cfg),
        None,
        |ctx, tx| {
            advisories::check(
                ctx,
                &dbs,
                lock,
                Option::<advisories::NoneReporter>::None,
                tx,
            );
        },
    )
    .unwrap();

    let yanked = [
        "quote 1.0.2 registry+https://github.com/rust-lang/crates.io-index",
        "spdx 0.3.1 registry+https://github.com/rust-lang/crates.io-index",
    ];

    for yanked in &yanked {
        assert!(
            diags.iter().any(|v| {
                field_eq!(v, "/fields/severity", "error")
                    && field_eq!(v, "/fields/message", "detected yanked crate")
                    && field_eq!(v, "/fields/labels/0/span", yanked)
            }),
            "failed to find yanked diagnostic for '{}'",
            yanked
        );
    }
}
