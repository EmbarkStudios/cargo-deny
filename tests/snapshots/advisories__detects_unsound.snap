---
source: tests/advisories.rs
expression: unsound_diag
---
{
  "fields": {
    "advisory": {
      "aliases": [
        "CVE-2019-25010",
        "CVE-2020-25575",
        "GHSA-jq66-xh47-j9f3",
        "GHSA-r98r-j25q-rmpr"
      ],
      "categories": [],
      "collection": "crates",
      "cvss": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H",
      "date": "2019-11-13",
      "description": "Safe Rust code can implement malfunctioning `__private_get_type_id__` and cause\ntype confusion when downcasting, which is an undefined behavior.\n\nUsers who derive `Fail` trait are not affected.",
      "id": "RUSTSEC-2019-0036",
      "informational": "unsound",
      "keywords": [
        "unsound"
      ],
      "package": "failure",
      "references": [],
      "related": [],
      "source": null,
      "title": "Type confusion if __private_get_type_id__ is overridden",
      "url": "https://github.com/rust-lang-nursery/failure/issues/336",
      "withdrawn": null
    },
    "code": "unsound",
    "graphs": [
      {
        "Krate": {
          "name": "failure",
          "version": "0.1.8"
        },
        "parents": [
          {
            "Krate": {
              "name": "advisories",
              "version": "0.1.0"
            }
          }
        ]
      }
    ],
    "labels": [
      {
        "column": 1,
        "line": 30,
        "message": "unsound advisory detected",
        "span": "failure 0.1.8 registry+https://github.com/rust-lang/crates.io-index"
      }
    ],
    "message": "Type confusion if __private_get_type_id__ is overridden",
    "notes": [
      "ID: RUSTSEC-2019-0036",
      "Advisory: https://rustsec.org/advisories/RUSTSEC-2019-0036",
      "Safe Rust code can implement malfunctioning `__private_get_type_id__` and cause\ntype confusion when downcasting, which is an undefined behavior.\n\nUsers who derive `Fail` trait are not affected.",
      "Announcement: https://github.com/rust-lang-nursery/failure/issues/336",
      "Solution: No safe upgrade is available!"
    ],
    "severity": "warning"
  },
  "type": "diagnostic"
}
